### DISCUSSION TITLE
Is it possible to use flux with an API as source instead of a repo/bucket?

### DISCUSSION THREAD

**Discussion by karakanb** [fluxcd/flux2#3556] [Q&A]
Hi there, my usecase is as follows:
- I have a system that has a concept called "client", created through a user interface. For each of these "client" definitions, I need to run a pod in my cluster with the client-specific configuration.
- As of today, I have a simple helm chart that generates a bunch of deployment based on the client list in the values file.
- Since the Helm chart is updated manually, every time there's a new client being created I need to go and change the values file and then deploy it.
- I'd like to automate this in a way that the corresponding deployment would be automatically created once a "client" is created in my system.
- I can expose the corresponding client list, or directly generate k8s manifests from my application, doesn't matter.
- If I can do this, that means I avoid maintaining a Git repo and adding things manually to the values file: instead, every time a new resource is created in my system, Flux would be able to create a deployment with the correct configuration.

It seems like Flux would be able to sync and deploy these changes if these were happening in a repo, however, I couldn't see if it would also be able to do it from an API endpoint that returns the data in whatever format. Is it possible?

**Comment by karakanb** [comment]
does anyone know if it is possible?

**Reply by wibed** [reply]
how would you approach this, besides the automation of the manual approach?

id imagine you can use `flux create ...` but i think only for flux resources right?

**Reply by karakanb** [reply]
what I'd imagine is a source as an API instead of an s3 bucket, so that when you make a `GET` call to the endpoint you get a list of templates and flux ensures the cluster is in sync with those templates. does that make sense?

**Reply by wibed** [reply]
well it seems odd to be so vague.

adding a file, and push it to git repo?
would flux automatically reconcile and create the resources for you?

**Reply by makkes** [reply]
This would really only make sense if the API provided a way for sources to be versioned and immutable as required by [the OpenGitOps manifest](https://opengitops.dev/) and from what I know the only APIs that provide these constraints are S3, Git and OCI. Basically your API needed to implement content addressing.

**Reply by karakanb** [reply]
I didn't intend to be vague actually, let me try to be a bit more explicit.

Imagine a table like this:
|client_id|image       |
|---------|------------|
|1        | python:3.10|
|2        | python:3.9 |

for each of the rows in this table, I would like to have a pod running in my cluster with the following deployment setup:
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: client-app
    client_id: "<client-id-here>"
spec:
  replicas: 1
  selector:
    matchLabels:
      app: client-app
      client_id: "<client-id-here>"
  template:
    metadata:
      labels:
        client_id: "<client-id-here>"
    spec:
      containers:
      - name: base
        image: <image-here>
```

the way I am doing this currently is to create these deployments manually via helm, but I'd like to automate this process and that's why I started looking into flux.

I have a few ways to automate this:
- interact with the kubernetes API directly from my application, send the deployment configuration every time a client row is created, updated or deleted.
    - this gets messy very easily, I'd like to avoid that if possible and also not give my application the ability to create/delete/update deployments if I can avoid it for security purposes.
- use flux with a Git repo & integrate my application to run Git commands: e.g. clone the repo, add the file, commit and push. this is doable, but it means I'd need to build the bits around integrating Git with my application, committing and pushing, handling errors there, etc.. Not impossible, but also not trivial, I'd rather avoid if I can.
- use flux with S3: instead of committing files to a Git repo and pushing them, I could put the files into the S3 bucket and eliminate some steps around Git, slightly simpler.

While thinking about these options another thought popped up in my mind: if flux was able to use a given API endpoint instead of an S3 bucket I could build a very simple API endpoint that looped through the rows in the table and return the YAML configs which would simplify the job quite a bit since I wouldn't need to integrate with Git or S3 or the kubernetes API. This is the primary reason why I created this discussion, not because I think it is a problem with flux or anything, just to brainstorm.

---

About the immutability, I haven't thought about that before. does s3 provide any versioning that cannot be removed actually? because I presume if the bucket is not versioned all we get is the latest version of the bucket, hence no way to go back. I might be wrong here, so I'd be happy if you could share some pointers so that I can learn a bit more about that.

In the meantime if this doesn't make much sense please feel free to close the discussion.

**Reply by wibed** [reply]
that makes more sense.

just as a thought. you could integrate a psql pub sub as a job runner that creates the necessary templates and git commands
https://webapp.io/blog/postgres-is-the-answer/

youd have to catalogue metadata anyways.

**Comment by makkes** [comment]
> if flux was able to use a given API endpoint instead of an S3 bucket I could build a very simple API endpoint that looped through the rows in the table and return the YAML configs which would simplify the job quite a bit since I wouldn't need to integrate with Git or S3 or the kubernetes API

So instead of using existing solutions like S3 you would rather implement your own solution, maintain the code and the supply chain behind it as well as roll it out? Why? You already mentioned Helm so why not just store the chart you're deploying somewhere (S3 or (better) OCI) and use proper values.yaml files to let Flux create releases based on this?

**Reply by karakanb** [reply]
I could do that as well, obviously. I don't intend to rebuild a solution to replace S3 or sth similar; however, I wanted to understand if there was a simpler way for a simpler use case. 

If I were to use S3, there are a few things I need to take care of:
- I need to set up background jobs and queue workers to be able to add new files/configurations on S3.
- I need to figure out the authentication between my application and S3 and store and maintain these credentials.
- I need to figure out a way to change existing files if I need to add anything new to the existing templates on S3, and possibly write some tooling for that.

I can do these as well, I am not saying it is impossible or too hard; however, I think it is way more complex than an API endpoint that can return a JSON or a YAML. There's no supply chain, no more maintenance than the code that'll interact with S3; a simple loop that runs over a SQL query and return an array. If there's a simpler solution I'd prefer that, if not that's also okay.

Of course, feel free to share in case I misunderstand the suggestion or if there's a simpler way.
